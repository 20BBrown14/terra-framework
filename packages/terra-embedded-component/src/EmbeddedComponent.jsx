/* global SystemJS */
import React from 'react';
import PropTypes from 'prop-types';
import 'systemjs/dist/system';
import { loadStylesheet } from './assets';

// Expose AMD required globals on the window
window.define = window.SystemJS.amdDefine;
window.require = window.requirejs = SystemJS.amdRequire;

const DefaultTimeout = () => <p>Component mounting is timed out.</p>;
const DefaultPlaceholder = () => <p>Loading</p>;

const propTypes = {

  /**
   * The basePath to load the manifest and manifest asset dependencies from.
   * */
  basePath: PropTypes.string.isRequired,

  /**
   * A placeholder to display while the component is loading.
   * */
  placeholder: PropTypes.element,

  /**
   * A component to display while the loading of css or module is timed out.
   * */
  timeoutComponent: PropTypes.element,

  /**
   * The number of milliseconds to wait before timing out.
   * @type {[type]}
   */
  timeoutInterval: PropTypes.number,

  /**
   * The entry file in the manifest to use to load the embeddable module.
   * */
  entry: PropTypes.string,

  /**
   * A manifest object defining the asset dependencies. If provided will be used instead
   * of fetching a manifest from the basePath.
   * */
  manifest: PropTypes.object,
};

const defaultProps = {
  entry: 'index.js',
  timeoutInterval: 3000,
  timeoutComponent: <DefaultTimeout />,
  placeholder: null,
};


class EmbeddedComponent extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      module: null,
      cssReady: false,
      timeout: false,
    };
  }

  componentDidMount() {
    this.fetchManifest().then((manifest) => {
      this.loadCSS(manifest);
      this.loadModule(this.getAbsolutePath(manifest[this.props.entry]));
    });

    // Sets a timer for timing out.
    setTimeout(() => {
      if (!this.state.module || !this.state.cssReady) {
        this.setState({ timeout: true });
      }
    }, this.props.timeoutInterval);
  }

  /**
  * Get the absolute path to the given filename.
  * @param {String} fileName - The file name to resolve to the absolute path.
  * @return {String} The absolute path to the given filename.
  */
  getAbsolutePath(fileName) {
    return `${this.props.basePath}/${fileName}`;
  }

  /**
  * Loads the CSS files detected within the manifest.
  * @param {object} manifest - The manifest used to load the CSS.
  */
  loadCSS(manifest) {
    const loadStylesheets = Object.keys(manifest)
      .filter(file => /\.css$/.test(file))
      .map(file => this.getAbsolutePath(manifest[file]))
      .map(loadStylesheet);

    Promise
      .all(loadStylesheets)
      .then(() => this.setState({ cssReady: true }));
  }

  /**
  * Fetches the asset manifest.
  * @return {Promise}
  */
  fetchManifest() {
    // If manifest is passed in, resolve immediately
    if (this.props.manifest) {
      return Promise.resolve(this.props.manifest);
    }

    // Fetch manifest from basePath
    return fetch(`${this.props.basePath}/manifest.json`, { cache: 'no-store' })
      .then(res => res.json());
  }

  /**
  * Loads the module specified by the url.
  * @param {String} url - The url to load
  */
  loadModule(url) {
    SystemJS.amdRequire([url], module => this.setState({ module }));
  }

  render() {
    const { module: Component, cssReady, timeout } = this.state;
    const { entry, manifest, placeholder, basePath, ...customProps } = this.props;

    if (timeout) {
      return this.props.timeoutComponent;
    }

    if (cssReady && Component) {
      return <Component {...customProps} />;
    }

    return this.props.placeholder;
  }
}

EmbeddedComponent.propTypes = propTypes;
EmbeddedComponent.defaultProps = defaultProps;

export default EmbeddedComponent;
