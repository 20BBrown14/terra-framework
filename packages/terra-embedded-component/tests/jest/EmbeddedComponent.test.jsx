import React from 'react';
import EmbeddedComponent from '../../src/EmbeddedComponent';

describe('EmbeddedComponent', () => {
  const basePath = './dist';
  const modulePath = 'index.js';
  const aModule = 'a module';
  const defaultRender = <EmbeddedComponent basePath={basePath} />;
  const withPlaceholder = <EmbeddedComponent basePath={basePath} placeholder={<div>placeholder</div>} />;
  const withTimeout = <EmbeddedComponent basePath={basePath} timeoutComponent={<div>timeout</div>} />;
  const aManifest = { 'index.js': modulePath, 'index.css': 'index.css' };
  const withManifest = <EmbeddedComponent basePath={basePath} manifest={aManifest} />;

  beforeEach(() => {
    // Mock Fetch
    global.fetch = jest.fn().mockImplementation(() => {
      const p = new Promise((resolve) => {
        resolve(
          {
            json: () => aManifest,
          },
        );
      });
      return p;
    });

    // Mock amdRequire
    global.SystemJS.amdRequire = jest.fn().mockImplementation((urlArray, callback) => {
      callback(aModule);
    });
  });

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should use blank message when no placeholder is given', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  it('should use the placeholder message when placeholder is given', () => {
    const wrapper = shallow(withPlaceholder);
    expect(wrapper).toMatchSnapshot();
  });

  it('should use the default timeout when no timeout is given', () => {
    const wrapper = shallow(defaultRender);
    wrapper.setState({ timeout: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should use the custom timeout when timeout is given', () => {
    const wrapper = shallow(withTimeout);
    wrapper.setState({ timeout: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should load the manifest and send to loadCSS when given', async () => {
    const wrapper = shallow(withManifest);
    // This mocks loadCSS function
    wrapper.instance().loadCSS = jest.fn((manifest) => {
      expect(manifest).toEqual(aManifest);
    });
    await wrapper.instance().componentDidMount();
  });

  it('should load the manifest through fetch when not given', async () => {
    const wrapper = shallow(defaultRender);
    // This mocks loadCSS function
    wrapper.instance().loadCSS = jest.fn((manifest) => {
      expect(manifest).toEqual(aManifest);
    });
    await wrapper.instance().componentDidMount();
  });

  it('should run loadModule with correct url', async () => {
    const wrapper = shallow(withManifest);
    // This mocks the loadModule function
    wrapper.instance().loadModule = jest.fn((url) => {
      expect(url).toEqual(`${basePath}/${modulePath}`);
    });
    await wrapper.instance().componentDidMount();
  });

  it('should load the module into state', async () => {
    const wrapper = shallow(withManifest);
    await wrapper.instance().componentDidMount();
    expect(wrapper.instance().state.module).toEqual(aModule);
  });
});
