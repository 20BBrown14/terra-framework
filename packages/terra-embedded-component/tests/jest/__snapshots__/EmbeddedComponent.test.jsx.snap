// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmbeddedComponent should render a default component 1`] = `null`;

exports[`EmbeddedComponent should use blank message when no placeholder is given 1`] = `null`;

exports[`EmbeddedComponent should use the custom timeout when timeout is given 1`] = `
<div>
  timeout
</div>
`;

exports[`EmbeddedComponent should use the default timeout when no timeout is given 1`] = `<DefaultTimeout />`;

exports[`EmbeddedComponent should use the placeholder message when placeholder is given 1`] = `
<div>
  placeholder
</div>
`;
